File: newton_cotes.jl
"""
Newton-Cotes 数値積分

閉じたニュートン・コーツ公式による数値積分の実装と可視化。
  - 台形則 (n=1)
  - Simpson 1/3 則 (n=2)
  - Simpson 3/8 則 (n=3)
  - Boole 則 (n=4)
"""

using LinearAlgebra
using Printf
using Plots

gr()  # PNG 保存に適した GR バックエンド

# ============================================================
# 単区間ルール
# ============================================================

trapezoid(f, a, b) = (b - a) / 2 * (f(a) + f(b))

simpson13(f, a, b) = (b - a) / 6 * (f(a) + 4f((a + b) / 2) + f(b))

function simpson38(f, a, b)
    h = (b - a) / 3
    (b - a) / 8 * (f(a) + 3f(a + h) + 3f(a + 2h) + f(b))
end

function boole(f, a, b)
    h = (b - a) / 4
    (b - a) / 90 * (7f(a) + 32f(a + h) + 12f(a + 2h) + 32f(a + 3h) + 7f(b))
end

# ============================================================
# 複合則（区間を N 等分して適用）
# ============================================================

function composite_trapezoid(f, a, b, N)
    xs = LinRange(a, b, N + 1)
    ys = f.(xs)
    h  = (b - a) / N
    h * (ys[1] / 2 + sum(ys[2:end-1]) + ys[end] / 2)
end

function composite_simpson13(f, a, b, N)
    N % 2 == 0 || throw(ArgumentError("N は偶数でなければなりません (N=$N)"))
    xs = LinRange(a, b, N + 1)
    ys = f.(xs)
    h  = (b - a) / N
    w  = fill(1.0, N + 1)
    w[2:2:end-1] .= 4
    w[3:2:end-2] .= 2
    h / 3 * dot(w, ys)
end

function composite_simpson38(f, a, b, N)
    N % 3 == 0 || throw(ArgumentError("N は 3 の倍数でなければなりません (N=$N)"))
    xs = LinRange(a, b, N + 1)
    ys = f.(xs)
    h  = (b - a) / N
    w  = fill(1.0, N + 1)
    for i in 2:N
        w[i] = (i - 1) % 3 == 0 ? 2.0 : 3.0
    end
    3h / 8 * dot(w, ys)
end

function composite_boole(f, a, b, N)
    N % 4 == 0 || throw(ArgumentError("N は 4 の倍数でなければなりません (N=$N)"))
    xs = LinRange(a, b, N + 1)
    ys = f.(xs)
    h  = (b - a) / N
    w  = zeros(N + 1)
    for k in 0:(N ÷ 4 - 1)
        for (j, wj) in enumerate([7, 32, 12, 32, 7])
            w[k * 4 + j] += wj
        end
    end
    2h / 45 * dot(w, ys)
end

# ============================================================
# 収束テーブル
# ============================================================

function print_convergence_table(f, a, b, exact, N_values)
    methods = [("Trapezoid",   composite_trapezoid),
               ("Simpson 1/3", composite_simpson13),
               ("Simpson 3/8", composite_simpson38),
               ("Boole",       composite_boole)]

    hdr = @sprintf("%-14s %5s %18s %12s %8s", "Method", "N", "Result", "Error", "Rate")
    println(hdr)
    println("-"^length(hdr))

    for (name, func) in methods
        prev_err = NaN
        for N in N_values
            result = try func(f, a, b, N) catch; continue end
            err    = abs(result - exact)
            if isnan(prev_err)
                @printf("%-14s %5d %18.12f %12.3e %8s\n", name, N, result, err, "  ---")
            else
                @printf("%-14s %5d %18.12f %12.3e %8.2f\n", name, N, result, err, log2(prev_err / err))
            end
            prev_err = err
        end
        println()
    end
end

# ============================================================
# ラグランジュ補間（可視化用）
# ============================================================

function lagrange_eval(xs, ys, t)
    n = length(xs)
    sum(ys[i] * prod((t - xs[j]) / (xs[i] - xs[j]) for j in 1:n if j ≠ i) for i in 1:n)
end

# ============================================================
# 図1: 各手法の近似の可視化（2×2 サブプロット）
# ============================================================

function plot_method_visualization(f, a, b; N=4, filename="fig_methods.png")
    x_fine = collect(LinRange(a, b, 500))
    y_fine = f.(x_fine)

    configs = [("Trapezoid Rule", 1, :steelblue),
               ("Simpson's 1/3",  2, :darkorange),
               ("Simpson's 3/8",  3, :forestgreen),
               ("Boole's Rule",   4, :crimson)]

    # 各手法の条件（N の倍数）に合わせて調整
    Ns_adj = [N, N, max(div(N, 3) * 3, 3), max(div(N, 4) * 4, 4)]

    subplots = map(zip(configs, Ns_adj)) do ((title, panel_size, color), n)
        x_nodes = collect(LinRange(a, b, n + 1))
        y_nodes = f.(x_nodes)

        p = plot(x_fine, y_fine; color=:black, lw=2, label="sin(x)",
                 title=title, xlabel="x", ylabel="f(x)",
                 xlims=(a, b), ylims=(-0.1, 1.25),
                 grid=true, gridalpha=0.3)

        for k in 0:(n ÷ panel_size - 1)
            idx      = (k * panel_size + 1):(k * panel_size + panel_size + 1)
            xs_panel = x_nodes[idx]
            ys_panel = y_nodes[idx]
            x_fill   = collect(LinRange(xs_panel[1], xs_panel[end], 100))
            y_poly   = lagrange_eval.(Ref(xs_panel), Ref(ys_panel), x_fill)
            plot!(p, x_fill, y_poly;
                  fill=0, fillalpha=0.3, fillcolor=color,
                  color=color, lw=1.2, label=false)
        end

        scatter!(p, x_nodes, y_nodes; color=color, ms=5, label="nodes (N=$n)")
        p
    end

    fig = plot(subplots..., layout=(2, 2), size=(900, 700))
    savefig(fig, filename)
    println("  -> $filename を保存しました")
end

# ============================================================
# 図2: 収束比較（log-log プロット）
# ============================================================

function plot_convergence(f, a, b, exact;
                          N_values=[12, 24, 48, 96, 192, 384],
                          filename="fig_convergence.png")
    methods = [("Trapezoid O(h²)",   composite_trapezoid, :steelblue,   :dash),
               ("Simpson 1/3 O(h⁴)", composite_simpson13, :darkorange,  :solid),
               ("Simpson 3/8 O(h⁴)", composite_simpson38, :forestgreen, :dashdot),
               ("Boole O(h⁶)",       composite_boole,     :crimson,     :dot)]

    p = plot(; xscale=:log10, yscale=:log10,
               xlabel="h = (b-a)/N", ylabel="Absolute Error",
               title="Newton-Cotes Convergence Comparison",
               legend=:bottomright, grid=true, gridalpha=0.3, size=(700, 550))

    all_hs = Float64[]; all_errs = Float64[]

    for (name, func, color, ls) in methods
        hs = Float64[]; errs = Float64[]
        for N in N_values
            result = try func(f, a, b, N) catch; continue end
            err = abs(result - exact)
            err > 0 && (push!(hs, (b - a) / N); push!(errs, err))
        end
        isempty(hs) && continue
        plot!(p, hs, errs; marker=:circle, color=color, linestyle=ls, lw=1.8, label=name)
        append!(all_hs, hs); append!(all_errs, errs)
    end

    # 参照線（O(h²), O(h⁴), O(h⁶)）
    if !isempty(all_hs)
        h_range = [minimum(all_hs), maximum(all_hs)]
        e_top   = maximum(all_errs)
        h_top   = maximum(all_hs)
        for order in [2, 4, 6]
            scale = e_top / h_top^order
            plot!(p, h_range, scale .* h_range .^ order;
                  color=:gray, lw=1, linestyle=:dash, alpha=0.5, label=false)
        end
    end

    savefig(p, filename)
    println("  -> $filename を保存しました")
end

# ============================================================
# メイン
# ============================================================

function main()
    f     = sin
    a, b  = 0.0, π
    exact = 2.0

    println("="^60)
    println("  Newton-Cotes 数値積分: sin(x) on [0, π]  (厳密値 = 2)")
    println("="^60)

    println("\n【単区間ルール（N=1）】")
    for (name, func) in [("台形則",       trapezoid),
                          ("Simpson 1/3", simpson13),
                          ("Simpson 3/8", simpson38),
                          ("Boole 則",    boole)]
        result = func(f, a, b)
        @printf("  %-14s: %.10f   誤差 = %.3e\n", name, result, abs(result - exact))
    end

    println("\n【複合則の収束テーブル】")
    print_convergence_table(f, a, b, exact, [12, 24, 48, 96, 192])

    println("【図の作成】")
    plot_method_visualization(f, a, b; N=4)
    plot_convergence(f, a, b, exact)
end

main()

File: newton_cotes.py
"""
Newton-Cotes 数値積分

閉じたニュートン・コーツ公式による数値積分の実装。
  - 台形則 (n=1)
  - Simpson 1/3 則 (n=2)
  - Simpson 3/8 則 (n=3)
  - Boole 則 (n=4)
各単区間ルールと、区間を N 等分して適用する複合則を含む。
"""

import numpy as np
import matplotlib
matplotlib.use("Agg")   # 画面なし環境でも PNG 保存できるようにする
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches


# ============================================================
# 単区間ルール
# ============================================================

def trapezoid(f, a: float, b: float) -> float:
    """台形則: [a, b] を 1 区間として積分（2 点）。
    次数 1 以下の多項式に対して厳密。
    誤差: -(b-a)^3/12 * f''(xi)
    """
    h = b - a
    return h / 2 * (f(a) + f(b))


def simpson13(f, a: float, b: float) -> float:
    """Simpson 1/3 則: [a, b] を 2 等分した 3 点を使用。
    次数 3 以下の多項式に対して厳密（超収束）。
    誤差: -(b-a)^5/90 * f^(4)(xi)
    """
    h = b - a
    m = (a + b) / 2
    return h / 6 * (f(a) + 4 * f(m) + f(b))


def simpson38(f, a: float, b: float) -> float:
    """Simpson 3/8 則: [a, b] を 3 等分した 4 点を使用。
    次数 3 以下の多項式に対して厳密。
    誤差: -(b-a)^5/80 * f^(4)(xi)
    """
    h = (b - a) / 3
    x1 = a + h
    x2 = a + 2 * h
    return (b - a) / 8 * (f(a) + 3 * f(x1) + 3 * f(x2) + f(b))


def boole(f, a: float, b: float) -> float:
    """Boole 則: [a, b] を 4 等分した 5 点を使用。
    次数 5 以下の多項式に対して厳密（超収束）。
    誤差: -2(b-a)^7/945 * f^(6)(xi)
    """
    h = (b - a) / 4
    x = [a + i * h for i in range(5)]
    w = [7, 32, 12, 32, 7]
    return (b - a) / 90 * sum(wi * f(xi) for wi, xi in zip(w, x))


# ============================================================
# 複合則（区間を N 等分して適用）
# ============================================================

def composite_trapezoid(f, a: float, b: float, N: int) -> float:
    """複合台形則: [a, b] を N 等分。収束次数 O(h^2), h=(b-a)/N。"""
    x = np.linspace(a, b, N + 1)
    y = f(x)
    h = (b - a) / N
    return h * (y[0] / 2 + np.sum(y[1:-1]) + y[-1] / 2)


def composite_simpson13(f, a: float, b: float, N: int) -> float:
    """複合 Simpson 1/3 則: N は偶数が必要。収束次数 O(h^4)。"""
    if N % 2 != 0:
        raise ValueError(f"N は偶数でなければなりません（N={N}）")
    x = np.linspace(a, b, N + 1)
    y = f(x)
    h = (b - a) / N
    # 重みパターン: 1, 4, 2, 4, 2, ..., 4, 1
    w = np.ones(N + 1)
    w[1:-1:2] = 4   # 奇数インデックス
    w[2:-2:2] = 2   # 偶数インデックス（両端以外）
    return h / 3 * np.dot(w, y)


def composite_simpson38(f, a: float, b: float, N: int) -> float:
    """複合 Simpson 3/8 則: N は 3 の倍数が必要。収束次数 O(h^4)。"""
    if N % 3 != 0:
        raise ValueError(f"N は 3 の倍数でなければなりません（N={N}）")
    x = np.linspace(a, b, N + 1)
    y = f(x)
    h = (b - a) / N
    # 重みパターン: 1, 3, 3, 2, 3, 3, 2, ..., 3, 3, 1
    w = np.ones(N + 1)
    for i in range(1, N):
        if i % 3 == 0:
            w[i] = 2
        else:
            w[i] = 3
    return 3 * h / 8 * np.dot(w, y)


def composite_boole(f, a: float, b: float, N: int) -> float:
    """複合 Boole 則: N は 4 の倍数が必要。収束次数 O(h^6)。"""
    if N % 4 != 0:
        raise ValueError(f"N は 4 の倍数でなければなりません（N={N}）")
    x = np.linspace(a, b, N + 1)
    y = f(x)
    h = (b - a) / N
    # 重みパターン（各パネル内）: 7, 32, 12, 32, 14, 32, 12, 32, 14, ..., 7
    w = np.zeros(N + 1)
    panel_w = [7, 32, 12, 32, 7]
    for k in range(N // 4):
        base = k * 4
        for j, wj in enumerate(panel_w):
            w[base + j] += wj
    return 2 * h / 45 * np.dot(w, y)


# ============================================================
# 図の作成
# ============================================================

def plot_method_visualization(f, a: float, b: float, N: int = 4, filename="fig_methods.png"):
    """各手法が関数をどう近似するかを可視化する（2×2 サブプロット）。"""
    x_fine = np.linspace(a, b, 500)
    y_fine = f(x_fine)

    methods = [
        ("台形則",       composite_trapezoid, "steelblue"),
        ("Simpson 1/3", composite_simpson13,  "darkorange"),
        ("Simpson 3/8", composite_simpson38,  "forestgreen"),
        ("Boole 則",    composite_boole,      "crimson"),
    ]
    # N が各手法の条件を満たすよう調整
    N_map = {"台形則": N, "Simpson 1/3": N, "Simpson 3/8": (N // 3) * 3 or 3, "Boole 則": (N // 4) * 4 or 4}

    fig, axes = plt.subplots(2, 2, figsize=(10, 7))
    fig.suptitle(f"Newton-Cotes 各手法の可視化  [sin(x) on [0, π], N={N}]", fontsize=13)

    for ax, (name, func, color) in zip(axes.flat, methods):
        n = N_map[name]
        x_nodes = np.linspace(a, b, n + 1)
        y_nodes = f(x_nodes)

        ax.plot(x_fine, y_fine, "k-", lw=2, label="sin(x)")

        # 各パネルを塗りつぶして近似を可視化
        panel_size = {"台形則": 1, "Simpson 1/3": 2, "Simpson 3/8": 3, "Boole 則": 4}[name]
        for k in range(n // panel_size):
            xs = x_nodes[k * panel_size: k * panel_size + panel_size + 1]
            ys = y_nodes[k * panel_size: k * panel_size + panel_size + 1]
            x_fill = np.linspace(xs[0], xs[-1], 200)
            # ラグランジュ補間で多項式を構築して塗りつぶす
            poly_coeffs = np.polyfit(xs, ys, len(xs) - 1)
            y_poly = np.polyval(poly_coeffs, x_fill)
            ax.fill_between(x_fill, 0, y_poly, alpha=0.3, color=color)
            ax.plot(x_fill, y_poly, color=color, lw=1.2)

        ax.plot(x_nodes, y_nodes, "o", color=color, ms=5, label=f"ノード (N={n})")
        ax.set_title(name, fontsize=11)
        ax.set_xlabel("x")
        ax.set_ylabel("f(x)")
        ax.legend(fontsize=8)
        ax.set_xlim(a, b)
        ax.grid(True, alpha=0.3)

    plt.tight_layout()
    plt.savefig(filename, dpi=150)
    print(f"  -> {filename} を保存しました")
    plt.close()


def plot_convergence(f, a: float, b: float, exact: float, N_values, filename="fig_convergence.png"):
    """各手法の収束を log-log プロットで比較する。"""
    funcs = {
        "台形則 O(h²)":       (composite_trapezoid, "steelblue",   "--"),
        "Simpson 1/3 O(h⁴)": (composite_simpson13,  "darkorange",  "-"),
        "Simpson 3/8 O(h⁴)": (composite_simpson38,  "forestgreen", "-."),
        "Boole 則 O(h⁶)":    (composite_boole,      "crimson",     ":"),
    }

    fig, ax = plt.subplots(figsize=(8, 6))

    for name, (func, color, ls) in funcs.items():
        hs, errs = [], []
        for N in N_values:
            try:
                result = func(f, a, b, N)
            except ValueError:
                continue
            err = abs(result - exact)
            if err > 0:
                hs.append((b - a) / N)
                errs.append(err)
        if hs:
            ax.loglog(hs, errs, marker="o", color=color, ls=ls, lw=1.8, label=name)

    # 理論収束次数の参照線
    h_ref = np.array([hs[0], hs[-1]])  # 最後のメソッドの h 範囲を流用
    for order, label, alpha in [(2, "∝h²", 0.4), (4, "∝h⁴", 0.4), (6, "∝h⁶", 0.4)]:
        scale = errs[0] * (hs[0] ** (-order)) if hs else 1
        ax.loglog(h_ref, scale * h_ref ** order, "gray", lw=1, alpha=alpha, ls="-")
        ax.text(h_ref[-1] * 1.05, scale * h_ref[-1] ** order, label, color="gray", fontsize=9)

    ax.set_xlabel("h = (b−a)/N", fontsize=12)
    ax.set_ylabel("絶対誤差", fontsize=12)
    ax.set_title("Newton-Cotes 各手法の収束比較", fontsize=13)
    ax.legend(fontsize=10)
    ax.grid(True, which="both", alpha=0.3)
    plt.tight_layout()
    plt.savefig(filename, dpi=150)
    print(f"  -> {filename} を保存しました")
    plt.close()


# ============================================================
# メイン: sin(x) を [0, π] で積分（厳密値 = 2）
# ============================================================

def _convergence_table(func_dict, f, a, b, exact, N_values):
    header = f"{'手法':<16} {'N':>5} {'結果':>18} {'誤差':>12} {'収束次数':>8}"
    print(header)
    print("-" * len(header))
    for name, func in func_dict.items():
        prev_err = None
        for N in N_values:
            try:
                result = func(f, a, b, N)
            except ValueError:
                continue
            err = abs(result - exact)
            rate = (np.log(prev_err / err) / np.log(2)) if prev_err and err > 0 else float("nan")
            prev_err = err
            print(f"{name:<16} {N:>5} {result:>18.12f} {err:>12.3e} {rate:>8.2f}")
        print()


if __name__ == "__main__":
    f = np.sin
    a, b = 0.0, np.pi
    exact = 2.0

    print("=" * 60)
    print("  Newton-Cotes 数値積分: sin(x) on [0, π]  (厳密値 = 2)")
    print("=" * 60)

    # --- 単区間ルールの比較 ---
    print("\n【単区間ルール（N=1）】")
    for name, func in [("台形則",        trapezoid),
                        ("Simpson 1/3",  simpson13),
                        ("Simpson 3/8",  simpson38),
                        ("Boole 則",     boole)]:
        result = func(f, a, b)
        print(f"  {name:<14}: {result:.10f}   誤差 = {abs(result - exact):.3e}")

    # --- 複合則の収束テーブル ---
    print("\n【複合則の収束テーブル】")
    funcs = {
        "台形則":       composite_trapezoid,
        "Simpson 1/3":  composite_simpson13,
        "Simpson 3/8":  composite_simpson38,
        "Boole 則":     composite_boole,
    }
    # LCM(1,2,3,4)=12 の倍数を使うと全手法で有効
    N_values = [12, 24, 48, 96, 192]
    _convergence_table(funcs, f, a, b, exact, N_values)

    # --- scipy との比較 ---
    try:
        from scipy.integrate import quad
        ref, _ = quad(f, a, b)
        print(f"scipy.integrate.quad 基準値: {ref:.15f}")
        result_boole = composite_boole(f, a, b, 32)
        print(f"複合 Boole (N=32):           {result_boole:.15f}  "
              f"差 = {abs(result_boole - ref):.3e}")
    except ImportError:
        print("（scipy がインストールされていないため比較をスキップ）")

    # --- 図の作成 ---
    print("\n【図の作成】")
    plot_method_visualization(f, a, b, N=4)
    plot_convergence(f, a, b, exact, N_values=[12, 24, 48, 96, 192, 384])
